### PR TITLE
Multivariable questions super PR

### PR DESCRIPTION
**Pull request by cameron-toy**
## What's New?
This PR combines several commits from previous PRs.

New:
- Support for multivariable questions! QA objects for questions involving multiple _properties_ can now be automatically generated and answered ([PROF]'s office hours are [PROF..OfficeHours] in [PROF..Office]).
- New Prof entity type. This entity replaces the existing Professors entity and is a combination of relevant information from the Professors, Sections, and OfficeHours entity. This unified entity allows answering many more fact-based questions than before.
- QA objects now take a database parameter during initialization, removing top-level initialization of the database in QA.py
- New QA functions for chaining database access, formatting professor office hours, and more.

Old (but still new):
- NLP performance increases.
- Removal of nltk and spacy as dependencies.
- Replaced NIMBUS_NLP.py with variable_extractor.py
- Added (currently unused) ProfessorSectionView entity.
- Replaced q_a_pairs.csv with @richagadgil's `get_all_answerable_pairs` function. Note: function is no longer used directly in question_classifier.py. See `build_question_classifier` for explanation.

## Type of change (pick-one)
- [ ] Bug fix (_non-breaking change which fixes an issue_)
- [x] New feature (_non-breaking change which adds functionality_)
- [ ] Breaking change (_fix or feature that would cause existing functionality to not work as expected_)
- [ ] This change requires a documentation update

## How Has This Been Tested?
See PRs #110 and #103. Also, multivariable question from database answered in if __name__==... block in nimbus.py.

Possible concerns:
- Slightly different overall features generated by spacy and spacy's stopwords list.
- Longer database access times for views.

TODO:
- Type annotation

## Checklist (check-all-before-merge)
_formatting help: `- [x]` means "checked' and `- [ ]` means "unchecked"_

- [ ] I documented my code according to the [Google Python Style Guide][1]

- [ ] I ran `./build_docs.sh` and the docs look fine

- [ ] I ran `./type_check.sh` and got no errors

- [x] I ran `./format.sh` because it automatically cleans my code for me 😄 

- [ ] I ran `./lint.sh` to check for what "format" missed

- [ ] I added my tests to the `/tests` directory

- [ ] I ran `./run_tests.sh` and all the tests pass

[1]: https://google.github.io/styleguide/pyguide.html

### CODE REVIEW

**Review by chidiewenike — APPROVED** [review]

File: Entity/Profs.py [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+    __tablename__ = "Profs"
+    firstName = Column(String(50), primary_key=True)
+    lastName = Column(String(50))
+    phoneNumber = Column(String(20))
+    researchInterests = Column(Text)
+    email = Column(String(255))
+    OfficeHours = Column(Text)
+    Platform = Column(Text)
+    LatestQuarter = Column(Text)
+    Office = Column(Text)
```
Comment (Jason-Ku) [inline Entity/Profs.py:18]: Are these supposed to be capitalized?

File: Entity/ProfessorSectionView.py [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+    seminar = Sem = 1
+    independent = Ind = 2
+    lab = 3
+    lecture = Lec = 4
+
+
+class ProfessorSectionView(Base):
+    __tablename__ = "Professor_Teaches_Section"
+    id = Column(Integer, primary_key=True)
+    id_sections = Column(Integer)
```
Comment (Jason-Ku) [inline Entity/ProfessorSectionView.py:20]: Looks like we're running into the issue of mixing camelCase and snake_case again... I'm sure there's a good reason to do it here but hopefully we can work something out in the future so we don't have to do this in a bunch of our Entity classes

File: nimbus_nlp/question_classifier.py [hunk 3]
```diff
@@ -1,287 +1,134 @@
...
+        doc = self.nlp(question)
+        test_features = self.get_question_features(doc)
+
+        array_gen = [
+            test_features[k] if k in test_features else 0 for k in self.overall_features
+        ]
+        test_array = np.array(array_gen)
+
+        # Flatten array into a vector
+        test_vector = test_array.reshape(1, -1)
```
Comment (Jason-Ku) [inline nimbus_nlp/question_classifier.py:119]: Nothing against reshape here, but I'm curious as to whether there's some benefit of using _reshape()_ here over _flatten()_

File: nimbus_nlp/question_classifier.py [hunk 4]
```diff
@@ -1,287 +1,134 @@
...
-        question_features = questions['features'].values.tolist()
-
-        # BUILD OVERALL FEATURE SET FROM INDIVIDUAL QUESTION FEATURE VECTORS
+        questions = [q[0] for q in question_pairs]
+        question_features = [self.get_question_features(self.nlp(q)) for q in questions]
         for feature in question_features:
             for key in feature:
-                if key not in self.overall_features:
-                    self.overall_features[key] = 0
+                self.overall_features[key] = 0
```
Comment (Jason-Ku) [inline nimbus_nlp/question_classifier.py:32]: nice catch on the optimization here

**Review by Jason-Ku — APPROVED** [review]

File: Entity/Profs.py (hunk 1 above)
Comment (chidiewenike) [reply Entity/Profs.py:18]: They match the field name on the SQL table.

**Review by snekiam — APPROVED** [review]
Looks good from a quick overview.  We should make a formatting check that must pass before a pr can be merged so we don't get huge formatting changelists.

**Comment by cameron-toy** [comment]
I’ll try to get a version of this PR passing style checks and type annotations up by this weekend.

**Pull request merged**